girder/utility/filesystem_assetstore_adapter.py

```diff
@@ -219,8 +219,9 @@ def finalizeUpload(self, upload, file):
 
         # Store the hash in the upload so that deleting a file won't delete
         # this file
-        upload['sha512'] = hash
-        self.model('upload').update({'_id': upload['_id']}, update={'$set': {'sha512': hash}})
+        if '_id' in upload:
```

Comment:
> was that we don't save an upload model when we import files, so not saving them when uploading would be okay.

The import logic and the upload logic go through different pathways in the assetstore code, so there's still risk in making this change that we break downstream implementations that were doing something like the filesystem assetstore was doing.

> We save the upload multiple times (which is part of why this is a significant speed up)

I do think it's a valuable optimization, and would love to have it now, but preserving behavior is more important IMO. My recommendation would be to open an issue and mark it with the **3.0** and **performance** labels. When we do release planning, it will go into the backlog of things to do prior to the release. I'd be OK with starting a 3.0-integration branch now if you'd like to push the code now rather than waiting.

> Perhaps wrongly, the user quota counts imported files the same as uploaded files.

It's nice that the user quota logic can be naive and just use the size totals -- since importing is an admin-only operation, we assume admins can also expand the destination quota as needed.